src/main/java/ch/njol/skript/patterns/PatternCompiler.java

```diff
@@ -0,0 +1,197 @@
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+/**
+ * The pattern structure is a linked list of {@link PatternElement}s,
+ * where {@link PatternElement#next} points to the next element to be matched,
+ * which can be on an outer level, and where {@link PatternElement#originalNext} points
+ * to the next element on the same level.
+ */
+public class PatternCompiler {
+
+	/**
+	 * @return an empty {@link PatternElement}
+	 */
+	private static PatternElement getEmpty() {
+		return new LiteralPatternElement("");
+	}
+
+	/**
+	 * Parses a pattern String into a {@link SkriptPattern}.
+	 */
+	public static SkriptPattern compile(String pattern) {
+		AtomicInteger atomicInteger = new AtomicInteger(0);
+		PatternElement first = compile(pattern, atomicInteger);
+		return new SkriptPattern(first, atomicInteger.get());
+	}
+
+	/**
+	 * Compiles the given pattern String into a pattern.<br>
+	 * The {@code expressionOffset} is to keep track of which index the next
+	 * {@link TypePatternElement} should be initiated with.
+	 * @return The first link of the {@link PatternElement} chain
+	 */
+	private static PatternElement compile(String pattern, AtomicInteger expressionOffset) {
+		StringBuilder literalBuilder = new StringBuilder();
+		PatternElement first = null;
+
+		for (int i = 0; i < pattern.length(); i++) {
+			char c = pattern.charAt(i);
+			if (c == '[') {
+				if (literalBuilder.length() != 0) {
+					first = appendElement(first, new LiteralPatternElement(literalBuilder.toString()));
+					literalBuilder = new StringBuilder();
+				}
+
+				int end = SkriptParser.nextBracket(pattern, ']', c, i + 1, true);
+				PatternElement patternElement = compile(pattern.substring(i + 1, end), expressionOffset);
+
+				first = appendElement(first, new OptionalPatternElement(patternElement));
+
+				i = end;
+			} else if (c == '(') {
+				if (literalBuilder.length() != 0) {
+					first = appendElement(first, new LiteralPatternElement(literalBuilder.toString()));
+					literalBuilder = new StringBuilder();
+				}
+
+				int end = SkriptParser.nextBracket(pattern, ')', c, i + 1, true);
+				PatternElement patternElement = compile(pattern.substring(i + 1, end), expressionOffset);
+
+				first = appendElement(first, new GroupPatternElement(patternElement));
+
+				i = end;
+			} else if (c == '|') {
+				if (literalBuilder.length() != 0) {
+					first = appendElement(first, new LiteralPatternElement(literalBuilder.toString()));
+					literalBuilder = new StringBuilder();
+				}
+
+				PatternElement prevFirst = first;
+
+				ChoicePatternElement choicePatternElement;
+				if (first instanceof ChoicePatternElement) {
+					choicePatternElement = (ChoicePatternElement) first;
+				} else {
+					first = choicePatternElement = new ChoicePatternElement();
+					choicePatternElement.add(prevFirst != null ? prevFirst : getEmpty());
+				}
+				choicePatternElement.add(getEmpty());
+			} else if (c == '¦') {
```

Comment:
What are your thoughts on adding ":" as an option for parse marks? This is a change skript-parser made that I think would be beneficial for us to introduce (at least as a secondary option).

Reply:
I think it'd be fine, but we do need to check if there are addons that use that character in syntax. It may be good to open an issue / discussion to get some feedback on this as well

Reply:
https://github.com/SkriptLang/Skript/discussions/4177